src/utils.cpp

```diff
@@ -246,6 +246,104 @@ bool UnsatCoreReducer::reduce_assump_unsatcore(const Term &formula,
   return true;
 }
 
+bool UnsatCoreReducer::linear_reduce_assump_unsatcore(
+                              const smt::Term &formula,
+                              const smt::TermVec &assump,
+                              smt::TermVec &out_red,
+                              smt::TermVec *out_rem,
+                              unsigned iter)
+{
+  TermVec bool_assump, tmp_assump;
+  UnorderedTermMap to_ext_assump;
+  TermVec cand_res;
+  for (const auto & a : assump) {
+    Term t = to_reducer_.transfer_term(a);
+    cand_res.push_back(t);
+    to_ext_assump[t] = a;
+  }
+
+  reducer_->push();
+  reducer_->assert_formula(to_reducer_.transfer_term(formula));
+
+  // exit if the formula is unsat without assumptions.
+  Result r = reducer_->check_sat();
+  if (r.is_unsat()) {
+    reducer_->pop();
+    return true;
+  }
+
+  for (const auto & a : cand_res) {
+    Term l = label(a);
+    reducer_->assert_formula(reducer_->make_term(Implies, l, a));
+    bool_assump.push_back(l);
+  }
+
+  unsigned cur_iter = 0;
+  unsigned assump_pos = 0;
+  r = reducer_->check_sat_assuming(bool_assump);
+  if (r.is_sat()) {
+    reducer_->pop();
+    return false;
+  }
+  assert(r.is_unsat());
+  TermVec bool_assump_query;
+  UnorderedTermSet core_set(bool_assump.begin(), bool_assump.end());
```

Comment:
do you actually need to populate the set here before the loop?

Reply:
This was for the case that not a single assumption can be removed. (In that case no UNSAT core extraction is executed)
But maybe I can write in another way that won't rely on the final `core_set`... let me change to that ...